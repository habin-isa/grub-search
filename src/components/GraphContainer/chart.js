import * as d3 from 'd3';

export const renderChart = (data, graphDiv) => {
  const height = 600;
  const width = 600;
  const links = data.links.map((d) => Object.create(d));
  const nodes = data.nodes.map((d) => Object.create(d));

  const color = () => {
    const scale = d3.scaleOrdinal(d3.schemeCategory10);
    console.log('puppi', (d) => scale(d.group));
    return (d) => scale(d.group);
  };

  const simulation = d3
    .forceSimulation(nodes)
    .force(
      'link',
      d3.forceLink(links).id((d) => d.id)
    )
    .force('link', d3.forceLink(links).distance(100))
    .force('charge', d3.forceManyBody())
    .force('center', d3.forceCenter(width / 2, height / 2));

  const drag = (simulation) => {
    const dragstarted = (d) => {
      if (!d3.event.active) simulation.alphaTarget(0.3).restart();
      d.fx = d.x;
      d.fy = d.y;
    };

    const dragged = (d) => {
      d.fx = d3.event.x;
      d.fy = d3.event.y;
    };

    const dragended = (d) => {
      if (!d3.event.active) simulation.alphaTarget(0);
      d.fx = null;
      d.fy = null;
    };

    return d3
      .drag()
      .on('start', dragstarted)
      .on('drag', dragged)
      .on('end', dragended);
  };

  const svg = graphDiv.append('svg').attr('viewBox', [0, 0, width, height]);

  const link = svg
    .append('g')
    .attr('stroke', '#999')
    .attr('strokeOpacity', 0.6)
    .selectAll('line')
    .data(links)
    .join('line')
    // .attr('strokeWidth', (d) => Math.sqrt(d.value));
    .attr('strokeWidth', 500);

  const node = svg
    .append('g')
    .attr('stroke', '#fff')
    .attr('strokeWidth', 1.5)
    .selectAll('circle')
    .data(nodes)
    .join('circle')
    .attr('r', 10)
    .attr('fill', color)
    .call(drag(simulation));

  node.append('title').text((d) => d.id);

  simulation.on('tick', () => {
    link
      .attr('x1', (d) => d.source.x)
      .attr('y1', (d) => d.source.y)
      .attr('x2', (d) => d.target.x)
      .attr('y2', (d) => d.target.y);

    node.attr('cx', (d) => d.x).attr('cy', (d) => d.y);
  });

  // invalidation.then(() => simulation.stop());
  console.log('Wanted elements:', svg);
  return svg.node();
};
