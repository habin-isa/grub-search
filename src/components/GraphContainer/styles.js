import styled from 'styled-components';

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 100%;
`;

export const Title = styled.div`
  font-size: 20px;
  font-weight: bold;
  color: hotpink;
`;
