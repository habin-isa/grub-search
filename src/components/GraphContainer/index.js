import React, { useState, useEffect, createRef } from 'react';
import * as S from './styles';
import { modifiedChart } from './chart';
import * as d3 from 'd3';
import * as d3Require from 'd3-require';
import { array, number } from 'prop-types';

const GraphContainer = ({ similarVenues, stopChart, seedVenue }) => {
  const [nodes, setNodes] = useState([]);
  const [links, setLinks] = useState([]);

  const data = {
    nodes: [],
    links: []
  };

  const chartInput = createRef();

  useEffect(
    () => {
      const myNode = document.getElementById('chartId');
      myNode.innerHTML = '';
      d3Require.require('d3@5');

      if (similarVenues.length > 0) {
        data.nodes.push({
          id: seedVenue.name,
          group: seedVenue.name,
          key: seedVenue.id,
          size: 20
        });
        {
          similarVenues.map((venue, i) => {
            data.nodes.push({
              id: venue.name,
              group: venue.name,
              key: venue.id,
              size: 20
            });
          });
          setNodes(data.nodes);
          console.log({ nodes });
        }
        const length = data.nodes.length;
        if (length > 1) {
          for (var i = 1; i < length; i++) {
            data.links.push({
              source: data.nodes[0].id,
              target: data.nodes[i].id,
              value: 3
            });
          }
          setLinks(data.links);
          console.log({ links });
          const chartData = modifiedChart(data, d3.select('.chartData'), stopChart);
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [similarVenues]
  );

  return (
    <S.Wrapper>
      <div id="chartId" className="chartData" />
    </S.Wrapper>
  );
};

GraphContainer.propTypes = {
  similarVenues: array,
  stopChart: number
};

GraphContainer.defaultProps = {
  similarVenues: [],
  stopChart: 0
};

export default GraphContainer;
