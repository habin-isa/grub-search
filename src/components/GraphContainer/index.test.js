import React from 'react';
import { shallow } from 'enzyme';
import { render } from '@testing-library/react';
import GraphContainer from '.';

describe('GraphContainer', () => {
  const props = {
    similarVenues: [
      { id: 'Costa', group: '1' },
      { id: 'Costa', group: '1' },
      { id: 'Costa', group: '1' }
    ]
  };
  it('should render', () => {
    const { container } = render(<GraphContainer {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
