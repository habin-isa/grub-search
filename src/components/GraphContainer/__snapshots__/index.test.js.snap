// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphContainer should render 1`] = `
<div
  class="sc-AxjAm iIQfHV"
>
  <div
    class="chartData"
    id="chartId"
  >
    <svg
      viewBox="0,0,400,400"
    >
      <g
        transform="translate(200,200)"
      >
        <g
          stroke="#000"
          stroke-width="1.5"
        />
        <g
          stroke="#fff"
          stroke-width="1.5"
        />
      </g>
    </svg>
  </div>
</div>
`;
